Kevin/Noah blog post (#60)
mechmotum/mechmotum.github.io: content/jupyter-initial-galaxy.rst

JupyterTeam Progress Summer 2020
================================

> :date: 2020-10-01 00:00:00

Author: I wrote it as "October 10th 2020" which was when I started the document. The format is a little odd admittedly, but do you also want the date to be more like 10/13 ?

> :date: 2020-10-01 00:00:00

Reviewer: Yes, make it the 13th, we can merge before the 13th is over. You all are up late!!

> :date: 2020-10-01 00:00:00

Reviewer 2: I read it as October 1st so I suggested a date update.

images which are built by Binder on that node, dramatically increasing
loading speeds when executing code cells on LibreTexts. Pressing "run"
will now take no more than 20 seconds to connect to a Binder provided
kernel whereas before it often took minutes or longer due to the
inconsistent availability of mybinder.org. This drastic decrease in load
times makes the plugin much more accessible for both authors and
readers.

Galaxy Cluster
--------------

The primary objectives for the new Galaxy cluster are high availability
and maintainability. Part of how we achieve high avialability is by
using `Keepalived and
HAproxy <https://github.com/kubernetes/kubeadm/blob/master/docs/ha-considerations.md#options-for-software-load-balancing>`__.

Keepalived runs as a static kubernetes pod on each control-plane node
and it manages a virtual IP for the kube-apiserver that all of the
kubernetes worker nodes communicate with. If the control-plane node
currently holding this Keepalived virtual IP goes down, Keepalived will
pass the virtual IP on to one of our remaining control-plane nodes so
that the rest of the kubernetes cluster can still communicate with the
kube-apiserver.

> After Keepalived recieves a kube-apiserver request, HAproxy (also

Reviewer 2: ```suggestion
After Keepalived receives a kube-apiserver request, HAproxy (also
```

running as a static pod on each control-plane node) will load-balance
the request onto whichever control-plane node is chosen according to a
`round robin
algorithm <https://avinetworks.com/glossary/round-robin-load-balancing/>`__.

> If we did not use HAproxy, any kube-apiserver request recieved through

Reviewer 2: ```suggestion
If we did not use HAproxy, any kube-apiserver request received through
```

the Keepalived virtual IP would automatically be routed to that same
control-plane node, leaving the other control-plane nodes to just sit
there and do nothing. Load-balancing with HAproxy takes better advantage
of our available hardware and lowers the kube-apiserver load on any
given node.

To improve the maintainability of this cluster, we use
`puppet <https://puppet.com/docs/puppet/6.18/puppet_index.html>`__ to
setup kubernetes and administrate our entire bare-metal cluster.

> Althought there already exists a `puppet module for boostrapping

Reviewer 2: ```suggestion
Although there already exists a `puppet module for bootstrapping
```

kubernetes <https://github.com/puppetlabs/puppetlabs-kubernetes>`__, our

> high availability setup has specific demands which required a `new

Reviewer 2: If possible, maybe mention what the specific demands were?

Author: I think there I just meant that the official module doesn't install HAproxy and Keepalived. We can probably explain more in detail personally cuz I don't wanna write it all out lol

puppet module <https://github.com/LibreTexts/protogalaxy>`__ to be

> written. The Protogalaxy puppet module boostraps all of the necessary

Reviewer 2: ```suggestion
written. The Protogalaxy puppet module bootstraps all of the necessary
```

certificates <https://v1-18.docs.kubernetes.io/docs/tasks/administer-cluster/kubeadm/kubeadm-certs/>`__
expired. The reason for this was that our cluster was very behind on
Ubuntu and kubernetes upgrades; the kube-apiserver and kubelet were on
v1.12 while the latest release was v1.18 at the time. One of the first
things we did was to upgrade the flock cluster across all the nodes, and
they now sit at a neat v1.19 for kubelet/kube-apiserver and v18.04.5 for
Ubuntu. Furthermore, we established a policy of upgrading the cluster
every 4 months so that the certificates do not meet their yearly
expiration date as they did before.

Other improvements were made as well. Kubernetes
`cert-manager <https://cert-manager.io/docs/installation/kubernetes/>`__
has been upgraded to v1, and `helm <https://v3.helm.sh/>`__ has also
been migrated over to v3. Our grafana alerts setup has been recrafted so
that we no longer have to manually reinput our dashboards if the pod
goes down. We also enabled IPMI interfaces on all the nodes, affording
us remote adminstration tools in light of Covid-19.

Future Plans
------------

Our top priorities moving forward are to enhance the executable code

> cell features of LibreText and continue development on the Galaxy

Reviewer 2: ```suggestion
cell features of LibreTexts and continue development on the Galaxy
```

cluster. We would like to fix the current issues with ipywidgets and
other interactive plotting features currently exhibited by our CKEditor

> Binder plugin. To compliment this, we must bring cell-to-cell

Reviewer 2: ```suggestion
Binder plugin. To complement this, we must bring cell-to-cell
```
